M2.4(B) - Add type annotations(platform.auth + platform.search)

hardikkat24 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #NA
2. This PR does the following: Added type annotations to core.platform.auth and core.platform.search.

## Essential Checklist

- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface in various display sizes (mainly phone, tablet, and desktop display size) to demonstrate that the changes made in this PR work correctly.
If the changes in your PRs are autogenerated via a script and you cannot provide proof for the changes then please leave a comment "No proof of changes needed because {{Reason}}".
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

hardikkat24 committed:
add annotations to core/platform/transactions

hardikkat24 committed:
datastore

hardikkat24 committed:
datastore test

hardikkat24 committed:
models and tests

hardikkat24 committed:
base_model

hardikkat24 committed:
activity model

hardikkat24 committed:
app_feedback_report model

hardikkat24 committed:
audit model

hardikkat24 committed:
auth model

hardikkat24 committed:
lint issues fixed

hardikkat24 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into m2.1

hardikkat24 committed:
lint

hardikkat24 committed:
lint

hardikkat24 committed:
mypy fixed

hardikkat24 committed:
lint

hardikkat24 committed:
fixes

hardikkat24 committed:
[] to get

hardikkat24 committed:
fixes type annotations

hardikkat24 committed:
datastore test

hardikkat24 committed:
models test

hardikkat24 committed:
activity model test

hardikkat24 committed:
base_models test

hardikkat24 committed:
added init to storage

hardikkat24 committed:
app_feedback_report tests

hardikkat24 committed:
audit tests

hardikkat24 committed:
auth tests

hardikkat24 committed:
lint fixed and files removed from denylist

hardikkat24 committed:
lint fixed

hardikkat24 committed:
merge

hardikkat24 committed:
no-implicit-reexport removed from config file

hardikkat24 committed:
lint+mypy fixed

hardikkat24 committed:
mock collection fixed

hardikkat24 committed:
ENUM Changed

hardikkat24 committed:
stubs added

hardikkat24 committed:
removed __init__

hardikkat24 committed:
lint checks

hardikkat24 committed:
email folder

hardikkat24 committed:
exploration models

hardikkat24 committed:
version int

hardikkat24 committed:
assert

hardikkat24 committed:
annotation multiple line

hardikkat24 committed:
else removed

hardikkat24 committed:
lint issues

hardikkat24 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into m2.1

hardikkat24 committed:
merge changes

hardikkat24 committed:
changes

hardikkat24 committed:
feedback partial

hardikkat24 committed:
query methods typed

hardikkat24 committed:
improved type annotations

hardikkat24 committed:
changes

hardikkat24 committed:
half feedback + improvements

hardikkat24 committed:
improvements

hardikkat24 committed:
lint

hardikkat24 committed:
vendor + fixes

hardikkat24 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into m2.1

hardikkat24 committed:
fix

hardikkat24 committed:
till opportunity partial

hardikkat24 committed:
opportunity complete

hardikkat24 committed:
user partial

hardikkat24 committed:
merged

hardikkat24 committed:
fixed after merge

hardikkat24 committed:
user complete

hardikkat24 committed:
addressed comments-1

hardikkat24 committed:
merge conf

hardikkat24 committed:
merge conf

hardikkat24 committed:
lint

hardikkat24 committed:
comments addressed-2

hardikkat24 committed:
changes

hardikkat24 committed:
fix tests

hardikkat24 committed:
addressed comments

hardikkat24 committed:
changes

hardikkat24 committed:
added explanations

hardikkat24 committed:
added todos and enum changed

hardikkat24 committed:
lint

hardikkat24 committed:
questions

hardikkat24 committed:
recommendation

hardikkat24 committed:
status

hardikkat24 committed:
addressed comments

hardikkat24 committed:
assert and type: ignore

hardikkat24 committed:
user assert

hardikkat24 committed:
comments for Any

hardikkat24 committed:
cast

hardikkat24 committed:
statistics

hardikkat24 committed:
story

hardikkat24 committed:
subtopic

hardikkat24 committed:
suggestions

hardikkat24 committed:
comments-split annotation

hardikkat24 committed:
addressed

hardikkat24 committed:
comments

hardikkat24 committed:
break annotations

hardikkat24 committed:
lint

hardikkat24 committed:
topic

hardikkat24 committed:
translation

hardikkat24 committed:
changes

hardikkat24 committed:
assert reason added

hardikkat24 committed:
Any

hardikkat24 committed:
split

hardikkat24 committed:
fixes

hardikkat24 committed:
lint

hardikkat24 committed:
app_identity

hardikkat24 committed:
auth incomplete

hardikkat24 committed:
lint-1

hardikkat24 committed:
comments

hardikkat24 committed:
fixes

hardikkat24 committed:
changes

hardikkat24 committed:
lint-2

hardikkat24 committed:
changes

hardikkat24 committed:
merge

hardikkat24 committed:
docstring

hardikkat24 committed:
merge conf

hardikkat24 committed:
Merge branch 'm2.2' into m2.3

hardikkat24 committed:
changes

hardikkat24 committed:
lint

hardikkat24 committed:
merge conf

hardikkat24 committed:
todo added

hardikkat24 committed:
removed main_taskqueue

hardikkat24 committed:
lint + mypy fix

hardikkat24 committed:
del

hardikkat24 committed:
lint

hardikkat24 committed:
lint

hardikkat24 committed:
lint

hardikkat24 committed:
merge conf

hardikkat24 committed:
comments

hardikkat24 committed:
lint

hardikkat24 committed:
lint

hardikkat24 committed:
Dict params

hardikkat24 committed:
merge conf

hardikkat24 committed:
1 pass full typed

hardikkat24 committed:
py2 -> py3

hardikkat24 committed:
changes

hardikkat24 committed:
unstack

hardikkat24 committed:
fixes

hardikkat24 committed:
lint

hardikkat24 committed:
changes

hardikkat24 committed:
lint

hardikkat24 committed:
changes

hardikkat24 committed:
search

hardikkat24 committed:
lints

hardikkat24 commented:
@sajalasati @vojtechjelinek PTAL

seanlip commented:
Hi @hardikkat24   -- sorry, could you please merge from develop? There was a backend test issue in develop that affected the CI checks on your PR, but it should be fixed now.

Thanks!

sajalasati commented on core/platform/auth/firebase_auth_services.py:
```diff
@@ -65,14 +65,21 @@
 from firebase_admin import exceptions as firebase_exceptions
+import webapp2
+
+from typing import Any, Dict, List, Optional, Text # isort:skip
```
Can you please stop using `Text` here and elsewhere?

hardikkat24 replied:
Done

sajalasati commented on core/platform/auth/firebase_auth_services.py:
```diff
@@ -96,7 +103,10 @@ def establish_firebase_connection():
 
 
-def establish_auth_session(request, response):
+def establish_auth_session(
+        request: webapp2.Request,
+        response: webapp2.Response
+) -> None:
```
```suggestion
def establish_auth_session(
        request: webapp2.Request,
        response: webapp2.Response) -> None:
```

Prefer keeping closing parenthesis on the same line (unless the return type annotation causes it to exceed 80 chars). Ditto for other places too.

cc @vojtechjelinek, @DubeySandeep for comments on this. My suggestion is based on what we have been doing elsewhere in the codebase since we don't have strict rules around this?

@DubeySandeep do you think the current styling is wrong? Should the closing parenthesis be right indented by 4 spaces?

hardikkat24 replied:
@sajalasati In our codebase, we are consistent in placing the return type in next line.

vojtechjelinek replied:
Personally, I like the sad-face approach (some arguments why are in this [blog-post](https://lukasz.langa.pl/1d1a43c4-9c8a-4c5f-a366-7f22ce6a49fc/) except for the consistency argument). I think it is also useful because if the type is too long (`really_long_variable: ReallyLongType) -> ReallyLongType:`) then it makes sense to split it up before the `)` anyway so using it by default makes it more consistent across the codebase.

sajalasati replied:
Yes I agree with the advantages, and perhaps following a single rule will help. Thanks!

sajalasati commented on core/platform/auth/firebase_auth_services_test.py:
```diff
@@ -190,7 +208,11 @@ def delete_user(self, uid):
+            self,
+            uids: List[Text],
+            force_delete: bool = False
+    ) -> 'firebase_auth.BatchDeleteAccountsResponse':
```
Why do we need to put this in quotes?

hardikkat24 replied:
Due to some unknown reason, `firebase_auth.BatchDeleteAccountsResponse` is not recognized here and throws firebase_auth has no attribute BatchDeleteAccountsResponse. But [here](https://github.com/oppia/oppia/blob/7c70f79de3ae649ef3c6114ffbb67c6c66f186d8/core/platform/auth/firebase_auth_services_test.py#L825) it is detected. 
Even it is not available in [firebase_admin.auth](https://github.com/firebase/firebase-admin-python/blob/master/firebase_admin/auth.py) on their github repo.  But it is available in[ firebase_admin._user_mgt](https://github.com/firebase/firebase-admin-python/blob/0e35c9a6ce7bc4bd8034c5369e601cb440cff061/firebase_admin/_user_mgt.py#L436). But if I use this, linter throws accessing private variables is not allowed. So putting it in quotes seems the best option. Thanks!

vojtechjelinek replied:
Is this something that `from __future__ import annotations` would help with? (I don't want to add it now, it is only useful for the future reference.

hardikkat24 replied:
I think this will help out in future. Thanks!

sajalasati replied:
@hardikkat24 Can you share the error log here - it seems weird, because if we can use that function, and we also have stubs defined for it, what exactly is the problem then. @vojtechjelinek are you able to understand this behaviour?

vojtechjelinek replied:
Not really, I imagine it can be related to some internal structure of firebase_admin where some components are loaded dynamically. But I don't think we need to bother with this, I think the long-term solution is to introduce `from __future__ import annotations` or move to Python 3.8 (which should be fairly easy compared to the Python 2 to Python 3 migration).

hardikkat24 replied:
>  Can you share the error log here

No error during mypy stubs because we have stubs. But error while running backend test. I agree with what @vojtechjelinek said. Thanks!

hardikkat24 replied:
> move to Python 3.8

@vojtechjelinek Are we going to do this soon?

vojtechjelinek replied:
I don't know, we still need to finish the Python 3 migration (mainly Beam jobs) so there are other priorities. On the other hand, I don't really want to delay it so we should do it in the next few months. I also assume it will be quite simple and easy.

sajalasati commented on core/platform/auth/firebase_auth_services_test.py:
```diff
@@ -391,12 +434,17 @@ def verify_id_token(self, token):
+            self,
+            session_cookie: Text,
+            check_revoked: bool = False
+    ) -> Dict[Text, Any]:
```
Missing TODO/explanation here.

hardikkat24 replied:
Found that we use only str, bool, None for values in our codebase for claims. See [here](https://github.com/oppia/oppia/blob/7c70f79de3ae649ef3c6114ffbb67c6c66f186d8/core/platform/auth/firebase_auth_services.py#L581). When we see the structure of [auth_domain.AuthClaims](https://github.com/oppia/oppia/blob/7c70f79de3ae649ef3c6114ffbb67c6c66f186d8/core/domain/auth_domain.py#L51), we see that Dict's type should be `Dict[str, Optional[Union[str, bool]]]`

sajalasati replied:
@hardikkat24 Sorry I don't get how you infer from AuthClaims structure the type info here using the links you gave. Which set of functions are you looking at in this file which leads you to this conclusion?

hardikkat24 replied:
@sajalasati  The attributes of the domain class led me to this conclusion. See [here](https://github.com/oppia/oppia/blob/7c70f79de3ae649ef3c6114ffbb67c6c66f186d8/core/domain/auth_domain.py#L57).

sajalasati replied:
Looked at the usage properly, and understood now. Thanks!

sajalasati commented on core/platform/auth/firebase_auth_services_test.py:
```diff
@@ -637,21 +722,35 @@ def _encode_user_claims(self, uid):
+    def _decode_user_claims(
+            self,
+            encoded_claims: Text
+    ) -> Optional[Dict[Text, Any]]:
```
Missing TODO/explanation here.

hardikkat24 replied:
Removed Any from here. Reason explained in comment above.

sajalasati commented on core/platform/auth/firebase_auth_services_test.py:
```diff
@@ -637,21 +722,35 @@ def _encode_user_claims(self, uid):
         """
         try:
-            return json.loads(encoded_claims)
+            # Casting the result here because the type stubs for library 'json'
```
(Just a note - this looks good! We should use similar explanations elsewhere too (where we use ANY) so that the purpose is loud and clear to every developer)

hardikkat24 replied:
Tried to do this for other usages of Any.

sajalasati commented on core/platform/search/elastic_search_services.py:
```diff
@@ -52,7 +54,11 @@ def _create_index(index_name):
 
-def add_documents_to_index(documents, index_name):
+# In the type annotation below Dict[str, Any] is used for documents because
+# there are no constraints for a document dictionary.
```
Can we point to the documentation/stubs here also as we did for similar cases in your other PR? You have already added stubs for elasticsearch, why we can't use some type that they define - that would be the best way to handle it. 

(I myself know that elastic search documents can be any json object, but it would be good to point to some external/internal resource to make it clear it others too as to why we have this - what and why is this special case if we are defining the type for it ourselves)

hardikkat24 replied:
Done.

sajalasati commented on core/platform/search/elastic_search_services.py:
```diff
@@ -178,7 +193,9 @@ def search(
-    query_definition = {
+    # In the type annotation below Dict[str, Any] is used for documents because
+    # the query_definition is a dictionary having values of various types.
+    query_definition: Dict[str, Any] = {
```
Ditto for queries: These are simple json objects without strict structure, but will be good to point to a resource.

hardikkat24 replied:
Done.

sajalasati reviewed (changes requested):
Thanks @hardikkat24! Sorry for delay in reviewing. I reviewed just the code files (i.e. not the stubs). My comments are about 2 main things: stop using `Text`, and make sure every `Dict[str, Any]` has todo/explanation. Had a quick pass on the type annotations, looks good to me.

hardikkat24 committed:
comments

hardikkat24 commented:
@sajalasati PTAL. I have addressed your comments.

hardikkat24 commented:
@DubeySandeep @nithusha21 PTAL.

vojtechjelinek commented on core/platform/auth/firebase_auth_services_test.py:
```diff
@@ -19,6 +19,8 @@
 from __future__ import absolute_import
 from __future__ import unicode_literals
 
+import contextlib  # pylint: disable=unused-import
```
Should we move all MyPy only imports to the `if` block? I think it would be beneficial if we decide to automatically ignore stuff like this only in the `if` block.

hardikkat24 replied:
Added in if MYPY block.

vojtechjelinek commented on core/platform/auth/firebase_auth_services_test.py:
```diff
@@ -519,13 +581,16 @@ def assert_is_not_user_multi(self, uids):
-            batch_error_pattern=(None,), individual_error_pattern=(None,)):
+            batch_error_pattern: Tuple[Optional[Exception]] = (None,),
+            individual_error_pattern: Tuple[Optional[bool]] = (None,)
+    ) -> 'contextlib._GeneratorContextManager[None]':
```
Why this weird type?

hardikkat24 replied:
Error is thrown If I use `ContextManager[List[str]]` for `capture_logging`.:
```
"ContextManager[List[str]]" of "capture_logging" incompatible with return type "_GeneratorContextManager[Any]" in supertype "TestBase"  [override]

```

Reason: See the type stubs of [contextlib.contextmanager](https://github.com/python/typeshed/blob/ee487304d76c671aa353a15e34bc1102bffa2362/stdlib/contextlib.pyi#L37)

vojtechjelinek replied:
Wait, how is `capture_logging` related to this?

hardikkat24 replied:
This error is thrown if I use ContentManager[List[str]] for capture_logging. This means that the self.swap(..) returns `_GeneratorContextManager[...]`. So that's why I used them here?
Should I change everything except capture_logging to use typing.ContextManager?

hardikkat24 replied:
Changed all of these to `ContextManager` except capture_logging. Thanks!

vojtechjelinek commented on core/platform/auth/firebase_auth_services_test.py:
```diff
@@ -599,28 +676,42 @@ def mock_import_users_error(
-                predicate=lambda record_and_error: record_and_error[1] is None,
-                enumerated=True)
+            records_to_import, records_to_fail = cast(
+                Tuple[
+                    List[Tuple[
+                        int, Tuple[firebase_auth.ImportUserRecord, str]]],
+                    List[Tuple[
+                        int, Tuple[firebase_auth.ImportUserRecord, str]]]
+                ],
```
Should we create a type variable for this?

hardikkat24 replied:
Done

vojtechjelinek commented on core/platform/auth/firebase_auth_services_test.py:
```diff
@@ -549,32 +614,44 @@ def mock_delete_users_error(
-
-            uids_to_delete = [uid for _, (uid, _) in uids_to_delete]
+            uids_to_delete, uids_to_fail = cast(
+                Tuple[
+                    List[Tuple[int, Tuple[str, str]]],
+                    List[Tuple[int, Tuple[str, str]]]
+                ],
```
Should this be separated into a type variable?

hardikkat24 replied:
Done

vojtechjelinek commented on core/platform/auth/firebase_auth_services_test.py:
```diff
@@ -220,9 +245,12 @@ def delete_users(self, uids, force_delete=False):
-                uids, predicate=lambda uid: self._users_by_uid[uid].disabled,
-                enumerated=True)
+            disabled_uids, enabled_uids = cast(
+                Tuple[List[Tuple[int, str]], List[Tuple[int, str]]],
```
Should this be separated into a type variable?

hardikkat24 replied:
Done

vojtechjelinek commented on core/platform/auth/firebase_auth_services_test.py:
```diff
@@ -800,16 +916,19 @@ class FirebaseAuthServicesTestBase(test_utils.AppEngineTestBase):
-    def capture_logging(self, min_level=logging.INFO):
+    def capture_logging(
+            self,
+            min_level: int = logging.INFO
```
Isn't there some enum for it in `logging`?

hardikkat24 replied:
No it is not an enum. See the stubs for logging [here](https://github.com/python/typeshed/blob/ee487304d76c671aa353a15e34bc1102bffa2362/stdlib/logging/__init__.pyi#L217).

vojtechjelinek replied:
Btw, both these can fit one line.

hardikkat24 replied:
Done

vojtechjelinek commented on core/platform/auth/firebase_auth_services_test.py:
```diff
@@ -1292,43 +1431,52 @@ class DeleteAuthAssociationsTests(FirebaseAuthServicesTestBase):
-    def swap_delete_user_to_always_fail(self):
+    def swap_delete_user_to_always_fail(
+            self
+    ) -> 'contextlib._GeneratorContextManager[None]':
```
Again, why this weird type?

hardikkat24 replied:
Explained in the comment earlier.

vojtechjelinek commented on core/platform/search/elastic_search_services.py:
```diff
@@ -129,9 +138,21 @@ def clear_index(index_name):
+        index_name: str,
+        categories: List[str],
+        language_codes: List[str],
+        offset: Optional[str] = None,
```
Why is this `str`, in the code we transform it to `int` anyway.

hardikkat24 replied:
I think changing this is not enough. This is linked throughout our codebase and we use str for offset. 
See [here](https://github.com/oppia/oppia/blob/a4a3c2b7c673e928264919b595f51222ddd8e2c5/core/domain/search_services.py#L167).
We already have a todo for this. See [here](https://github.com/oppia/oppia/issues/11314).

vojtechjelinek replied:
Ah, okay.

vojtechjelinek commented on stubs/google/appengine/api/app_identity/__init__.pyi:
```diff
@@ -0,0 +1,5 @@
+
+def get_application_id() -> Text: ...
+
+def get_default_gcs_bucket_name(deadline: Optional[float] = ...) -> Text: ...
```
Why do we need this file? I don't think we use this library anymore.

hardikkat24 replied:
removed

vojtechjelinek commented on stubs/firebase_admin/auth.pyi:
```diff
@@ -0,0 +1,92 @@
+
+class UserRecord:
+    def __init__(
+            self,
+            data: Dict[str, Any]
```
```suggestion
            self, data: Dict[str, Any]
```

hardikkat24 replied:
Done

vojtechjelinek commented on stubs/firebase_admin/auth.pyi:
```diff
@@ -0,0 +1,92 @@
+
+def verify_session_cookie(
+        session_cookie: str,
+        check_revoked: bool = ...,
```
Why is the `= ..` part needed here?

hardikkat24 replied:
This `...` part means that a default value is specified in the function for this.

vojtechjelinek reviewed (changes requested):
Thanks! Took a full pass. One thing that I'm unsure about is that whether we need all the elastic stubs or if we should only add the ones that we need and if we later need more just add links where we can actually find the sources for these.

hardikkat24 committed:
changes

hardikkat24 committed:
comments

hardikkat24 commented:
@vojtechjelinek PTAL. I have addressed all your comments. 
What do you prefer?
1. Add all the stubs(size 527 kb)
    - We don’t need to worry about adding stubs in future. Stubs for all the methods/classes would already be present
    - Increase the size of codebase (we won’t be using all the stubs).
2. Add only the required stubs
    - Process of adding stubs could be complicated if we are going to modify platform.search.


I am also not sure as to how should we proceed. Both options seem good. Thanks!!

nithusha21 reviewed (approved):
LGTM for 2 codeowner files

vojtechjelinek commented on core/platform/auth/firebase_auth_services_test.py:
```diff
@@ -38,9 +39,34 @@
 auth_models, user_models = (
     models.Registry.import_models([models.NAMES.auth, models.NAMES.user]))
 
+UidsPartitionTupleType = (
+    Tuple[
+        List[Tuple[int, str]],
+        List[Tuple[int, str]]
+    ])
+
+UidsZipPartitionTupleType = (
+    Tuple[
+        List[Tuple[int, Tuple[str, str]]],
+        List[Tuple[int, Tuple[str, str]]]
+    ])
+
+RecordsPartitionTupleType = (
+    Tuple[
+        List[Tuple[int, Tuple[firebase_auth.ImportUserRecord, str]]],
+        List[Tuple[int, Tuple[firebase_auth.ImportUserRecord, str]]]
+    ])
```
Should we do this?
```suggestion
UidsPartitionTupleType = Tuple[
    List[Tuple[int, str]],
    List[Tuple[int, str]]
]

UidsZipPartitionTupleType = Tuple[
    List[Tuple[int, Tuple[str, str]]],
    List[Tuple[int, Tuple[str, str]]]
]

RecordsPartitionTupleType = Tuple[
    List[Tuple[int, Tuple[firebase_auth.ImportUserRecord, str]]],
    List[Tuple[int, Tuple[firebase_auth.ImportUserRecord, str]]]
]
```

hardikkat24 replied:
Done

vojtechjelinek commented on core/platform/auth/firebase_auth_services_test.py:
```diff
@@ -845,7 +978,10 @@ def create_request(self, id_token=None, session_cookie=None):
 
-    def create_response(self, session_cookie=None):
+    def create_response(
+            self,
+            session_cookie: Optional[str] = None
```
```suggestion
            self, session_cookie: Optional[str] = None
```

hardikkat24 replied:
Done

vojtechjelinek commented on core/platform/auth/firebase_auth_services_test.py:
```diff
@@ -702,7 +821,10 @@ def _create_list_users_page_fragile(self, page_list, page_index):
 
-    def _create_delete_users_result_fragile(self, errors):
+    def _create_delete_users_result_fragile(
+            self,
+            errors: List[Tuple[int, str]]
```
```suggestion
            self, errors: List[Tuple[int, str]]
```

hardikkat24 replied:
Done

vojtechjelinek commented on core/platform/auth/firebase_auth_services_test.py:
```diff
@@ -719,7 +841,11 @@ def _create_delete_users_result_fragile(self, errors):
 
-    def _create_user_import_result_fragile(self, total, errors):
+    def _create_user_import_result_fragile(
+            self,
+            total: int,
+            errors: List[Tuple[int, str]]
```
```suggestion
            self, total: int, errors: List[Tuple[int, str]]
```

hardikkat24 replied:
Done

vojtechjelinek commented on core/platform/auth/firebase_auth_services_test.py:
```diff
@@ -381,7 +445,10 @@ def update_user(self, uid, email=None, disabled=False, custom_claims=None):
 
-    def verify_id_token(self, token):
+    def verify_id_token(
+            self,
+            token: str
```
```suggestion
            self, token: str
```

hardikkat24 replied:
Done

vojtechjelinek commented on core/platform/auth/firebase_auth_services_test.py:
```diff
@@ -344,7 +398,11 @@ def revoke_refresh_tokens(self, uid):
 
-    def set_custom_user_claims(self, uid, custom_claims):
+    def set_custom_user_claims(
+            self,
+            uid: str,
+            custom_claims: Optional[str]
```
```suggestion
            self, uid: str, custom_claims: Optional[str]
```

hardikkat24 replied:
Done

vojtechjelinek commented on core/platform/auth/firebase_auth_services_test.py:
```diff
@@ -190,7 +229,11 @@ def delete_user(self, uid):
 
-    def delete_users(self, uids, force_delete=False):
+    def delete_users(
+            self,
+            uids: List[str],
+            force_delete: bool = False
```
```suggestion
            self, uids: List[str], force_delete: bool = False
```

hardikkat24 replied:
Done

vojtechjelinek reviewed (changes requested):
Thanks! Left some more comments. Regarding the stubs for elasticsearch I think we should prefer only the minimal subset.

hardikkat24 committed:
changes;

hardikkat24 committed:
stubs changed

hardikkat24 committed:
ContextManager changes

hardikkat24 committed:
lint

hardikkat24 commented:
@vojtechjelinek PTAL. I have addressed all your comments and used only the stubs that are used by us. Thanks!

DubeySandeep reviewed (approved):
LGTM for the codeowner file `scripts/run_mypy_checks.py`

vojtechjelinek commented on stubs/elasticsearch/client/__init__.pyi:
```diff
@@ -0,0 +1,210 @@
+        self, hosts: Any = ..., transport_class: Type[Transport] = ..., **kwargs: Any
+    ) -> None: ...
+    def __repr__(self) -> str: ...
+    def __enter__(self) -> "Elasticsearch": ...
```
```suggestion
    def __enter__(self) -> 'Elasticsearch': ...
```

hardikkat24 replied:
Done

vojtechjelinek reviewed (changes requested):
LGTM! Just one last nit.

hardikkat24 committed:
change

hardikkat24 committed:
merge conf

sajalasati reviewed (approved):
LGTM!
